grpc: add trace.EventLogs to ClientConn and Server

Sajmani opened the pull request:
golang.org/x/net/trace's Event logs keep a bounded list of text log entries in memory that are
accessible via the /debug/events HTTP endpoint.  The logs are organized by type, instance,
and time-since-last-error.  They provide a simple way to find objects with recent errors and
inspect their events when debugging live systems.  Event logs are suitable only for relatively
infrequent events: connection setup and teardown, not individual RPCs.

Change-Id: If01b58e1441f88627e084491e0956e4df700e3bc

iamqizhao commented on clientconn.go:
```diff
@@ -157,6 +159,7 @@ func Dial(target string, opts ...DialOption) (*ClientConn, error) {
 		// Start a goroutine connecting to the server asynchronously.
 		go func() {
 			if err := cc.resetTransport(false); err != nil {
+				cc.events.Errorf("dial failed: %v", err)
```
This should be applied to the sync connecting case at line 152 too.

Sajmani replied:
Done.

iamqizhao commented on clientconn.go:
```diff
@@ -411,6 +419,7 @@ func (cc *ClientConn) Close() error {
 	}
 	cc.state = Shutdown
 	cc.stateCV.Broadcast()
+	cc.events.Finish()
```
suggest using "defer" for this at line 416 because the exit path at line 418 also needs to call Finish.

Sajmani replied:
Finish should only be called once.  The return at line 418 happens when Close has already been called; we don't want to call Finish in that case.

iamqizhao commented:
my last comment: I prefer letting the connection level tracing is also guarded by EnableTracing var (https://github.com/grpc/grpc-go/blob/fe3b16a0f35bdda0a7f400dc4fdb8d9342463d0a/trace.go#L49) which is currently used for rpc level tracing. What do you think?

Sajmani commented:
I suppose it's your call whether EnableTracing governs event logging, but it's *much* lower overhead than request tracing, since it's not on the RPC critical path and happens for infrequent events.

iamqizhao commented:
Yep, it is not performance critical but I do prefer that to make the story of tracing consistent. Can you do that in this PR?

Sajmani commented:
OK!  I've managed to restore this pull request and commit the change to do event logging only when EnableTracing is true.  I believe this addresses your last comment.

iamqizhao commented on clientconn.go:
```diff
@@ -411,6 +433,9 @@ func (cc *ClientConn) Close() error {
 	cc.state = Shutdown
 	cc.stateCV.Broadcast()
+	if cc.events != nil {
+		cc.events.Finish()
```
set cc.events to nil to avoid further invoking?

iamqizhao replied:
What if cc.events.Printf is called after cc.event.Finish() has been called? Is it no-op or there is some issue? If it is an issue, we need to set cc.events to nil here but then we still need a mutex to protect cc.events.

I actually tend to put the operations on cc.events (Printf, errorf, finish ) along with the change of cc.state. cc.state records the state change of ClientConn (and here we just need to push it to tracing) and has been protected by a mutex.

iamqizhao replied:
is there any problem if cc.event.Errorf/Printf get called after Finish is called?

Sajmani replied:
Sorry, I'm still getting used to this UI.  I've set cc.events = nil here and s.events = nil in server.go, after Finish.

iamqizhao replied:
Yeah, github review UI is horrible ... :)

Don't we need a mutex to guard cc.events and s.events (because you reset its value to nil now)? I think it can align with cc.state to share the same lock then.

Sajmani replied:
Ah, I see. Now that we're setting it to nil under the mutex, we need to put all the accesses to events under the same mutex.  I'll make that change.

iamqizhao commented on clientconn.go:
```diff
@@ -206,6 +224,7 @@ type ClientConn struct {
 	authority    string
 	dopts        dialOptions
 	shutdownChan chan struct{}
+	events       trace.EventLog
```
I think events should be protected by "mu" to avoid race?

Sajmani replied:
EventLog has its own mutex; it's safe for concurrent access.

iamqizhao commented:
Sorry about the delay. was busy on wrapping up my intern's work last week.

iamqizhao commented on clientconn.go:
```diff
@@ -167,6 +173,18 @@ func Dial(target string, opts ...DialOption) (*ClientConn, error) {
 	return cc, nil
 }
 
+func (cc *ClientConn) printf(format string, a ...interface{}) {
```
rename to eventPrintf and eventErrorf respectively?

Sajmani replied:
Per chat, can we leave as is? Anything else needed for this PR?
On Aug 25, 2015 5:24 PM, "Qi Zhao" <notifications@github.com> wrote:

> In clientconn.go
> <https://github.com/grpc/grpc-go/pull/287#discussion_r37921644>:
>
> > @@ -167,6 +173,18 @@ func Dial(target string, opts ...DialOption) (*ClientConn, error) {
> >  	return cc, nil
> >  }
> >
> > +func (cc *ClientConn) printf(format string, a ...interface{}) {
>
> rename to eventPrintf and eventErrorf respectively?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/287/files#r37921644>.
>

iamqizhao commented:
My biggest concern is as follows (copy and paste from the previous comment you probably missed):

"I actually tend to put the operations on cc.events (Printf, errorf, finish ) along with the change of cc.state. cc.state records the state change of ClientConn (and here we just need to push it to tracing) and has been protected by a mutex."

It sounds a natural choice.

Sajmani commented:
If we need to read cc.state to compose the trace message, then yes, we
should hold the mutex.  But otherwise it's just holding the mutex
unnecessarily; it seems better to avoid holding it during operations that
don't need it.
On Sep 4, 2015 4:21 PM, "Qi Zhao" <notifications@github.com> wrote:

> My biggest concern is as follows (copy and paste from the previous comment
> you probably missed):
>
> "I actually tend to put the operations on cc.events (Printf, errorf,
> finish ) along with the change of cc.state. cc.state records the state
> change of ClientConn (and here we just need to push it to tracing) and has
> been protected by a mutex."
>
> It sounds a natural choice.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/287#issuecomment-137844088>.
>

iamqizhao commented on clientconn.go:
```diff
@@ -320,9 +339,11 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 			grpclog.Printf("grpc: ClientConn.resetTransport failed to create client transport: %v; Reconnecting to %q", err, cc.target)
 			continue
 		}
+		cc.printf("connection established")
```
move to line#363?

iamqizhao commented:
regardless mutex, my point is that the trace log should (almost) be aligned with cc.state change, e.g.,
cc.state = Ready --> log connection established;
cc.state = TransientFailure --> log connection retry;
cc.state = Shutdown --> finish trace;

It seems you are already doing this if you address my last comment about "connection established". So we are good to go once you fix that one. Thanks for the contribution.

Sajmani commented:
Ok, I've added a new commit that records a trace event for each client conn state change.

iamqizhao commented on server.go:
```diff
@@ -197,11 +217,13 @@ func (s *Server) Serve(lis net.Listener) error {
 			return err
 		}
 		if creds, ok := s.opts.creds.(credentials.TransportAuthenticator); ok {
 			c, err = creds.ServerHandshake(c)
```
This seems on an out-of-date version. See https://github.com/grpc/grpc-go/blob/master/server.go#L204. You probably need to sync your client.

iamqizhao commented:
i) The following comment did not get answered:
"is there any problem if cc.event.Errorf/Printf get called after cc.Event.Finish is called?"
ii) It seems you need to sync your git client and resolve conflicts.

Sajmani commented:
OK!  Still figuring out GitHub.  I think I've addressed your remaining comments; sorry about all the back-and-forth.

Sajmani commented:
Qi Zhao, I believe I've addressed all your comments. I'd like to present this feature when I speak on gRPC at GothamGo Oct 2. Can we get this merged soon? Thanks.

iamqizhao commented:
Sure, let's get this merged on next Monday (if not during this weekend). I
think the revision will take you 30 min at most. :-)

On Sat, Sep 19, 2015 at 4:08 AM, Sameer Ajmani <notifications@github.com>
wrote:

> Qi Zhao, I believe I've addressed all your comments. I'd like to present
> this feature when I speak on gRPC at GothamGo Oct 2. Can we get this merged
> soon? Thanks.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/grpc/grpc-go/pull/287#issuecomment-141650675>.
>

Sajmani commented:
OK, locking is all fixed.  PTAL.

iamqizhao commented on clientconn.go:
```diff
@@ -337,6 +365,7 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 		connectTime := time.Now()
 		newTransport, err := transport.NewClientTransport(cc.target, &copts)
 		if err != nil {
+			cc.errorf("transient failure: %v", err)
```
put it under cc.mu.Lock() (line#369).

Sajmani replied:
Done.

iamqizhao commented on clientconn.go:
```diff
@@ -347,6 +376,7 @@ func (cc *ClientConn) resetTransport(closeTransport bool) error {
 			}
 			// Fail early before falling into sleep.
 			if cc.dopts.copts.Timeout > 0 && cc.dopts.copts.Timeout < sleepTime+time.Since(start) {
+				cc.errorf("connection timeout")
```
guard by cc.mu?

Sajmani replied:
Done.

iamqizhao commented:
And there are some minor conflicts with the latest version. Please sync.
